File: Data Structures & Algorithms/insert-new-interval/submission-0.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        output = []
        for i in range(len(intervals)):
            # first case, new interval to the left of the current ith interval
            if newInterval[1] < intervals[i][0]:
                output.append(newInterval)
                output += intervals[i:]
                return output
            elif newInterval[0] > intervals[i][1]: # new interval completely to the right of ith interval, append all upto ith
                output.append(intervals[i])
            else:
                # overlap with current, so merge
                newInterval = [min(newInterval[0], intervals[i][0]), max(newInterval[1], intervals[i][1])]
        output.append(newInterval)
        return output
        

File: Data Structures & Algorithms/merge-intervals/submission-0.cpp
class Solution {
public:

    static bool compareIntervals(const vector<int>& i1, const vector<int>& i2) {
        return i1[0] <i2[0];
    }
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), compareIntervals);
        vector<vector<int>> result;
        result.push_back(intervals[0]);
        for (auto& interval: intervals) {
            // check if overlap
            if (interval[0] <= result.back()[1]) {
                result.back()[1] = max(result.back()[1], interval[1]);
            }
            else {
                result.push_back(interval);
            }
        }
        return result;
    }
    
};
